packages/assets-controllers/src/TokenDetectionController.test.ts

```diff
@@ -226,366 +162,946 @@ describe('TokenDetectionController', () => {
+      );
     });
 
-    tokensController.ignoreTokens([sampleTokenA.address]);
-
-    getBalancesInSingleCall.resolves({
-      [sampleTokenA.address]: new BN(1),
+    it('should not detect tokens if there is no selectedAddress set', async () => {
+      const mockGetBalancesInSingleCall = jest.fn().mockResolvedValue({
+        [sampleTokenA.address]: new BN(1),
+      });
+      const mockAddDetectedTokens = jest.fn();
+      await withController(
+        {
+          options: {
+            addDetectedTokens: mockAddDetectedTokens,
+            getBalancesInSingleCall: mockGetBalancesInSingleCall,
+            networkClientId: NetworkType.mainnet,
+            selectedAddress: '',
+          },
+        },
+        async ({ controller, mockTokenListGetState }) => {
+          mockTokenListGetState.mockReturnValue({
+            ...getDefaultTokenListState(),
+            tokenList: {
+              [sampleTokenA.address]: {
+                name: sampleTokenA.name as string,
+                symbol: sampleTokenA.symbol,
+                decimals: sampleTokenA.decimals,
+                address: sampleTokenA.address,
+                occurrences: 1,
+                aggregators: sampleTokenA.aggregators,
+                iconUrl: sampleTokenA.image,
+              },
+            },
+          });
+
+          await controller.start();
+
+          expect(mockAddDetectedTokens).not.toHaveBeenCalled();
+        },
+      );
     });
-    await tokenDetection.detectTokens();
-    expect(tokensController.state.tokens).toStrictEqual([sampleTokenB]);
-
-    expect(tokensController.state.ignoredTokens).toStrictEqual([
-      sampleTokenA.address,
-    ]);
   });
 
-  it('should add a token when detected with a balance even if it is ignored on another account', async () => {
-    preferences.setSelectedAddress('0x0001');
-    changeNetwork(mainnet);
-
-    await tokenDetection.start();
-
-    await tokensController.addToken({
-      address: sampleTokenA.address,
-      symbol: sampleTokenA.symbol,
-      decimals: sampleTokenA.decimals,
+  describe('PreferencesController:stateChange', () => {
+    let clock: sinon.SinonFakeTimers;
+    beforeEach(() => {
+      clock = sinon.useFakeTimers();
     });
 
-    tokensController.ignoreTokens([sampleTokenA.address]);
-
-    preferences.setSelectedAddress('0x0002');
-
-    getBalancesInSingleCall.resolves({
-      [sampleTokenA.address]: new BN(1),
+    afterEach(() => {
+      clock.restore();
     });
-    await tokenDetection.detectTokens();
-    expect(tokensController.state.detectedTokens).toStrictEqual([sampleTokenA]);
-  });
-
-  it('should not autodetect tokens that exist in the ignoreList', async () => {
-    preferences.update({ selectedAddress: '0x1' });
-    changeNetwork(mainnet);
 
-    await tokenDetection.start();
-
-    getBalancesInSingleCall.resolves({
-      [sampleTokenA.address]: new BN(1),
-    });
-    await tokenDetection.detectTokens();
+    describe('enabled', () => {
```

Comment:
What do you think about using:

``` suggestion
    describe('when the "disabled" option is false', () => {
```

or something like that?

Reply:
Good point, done in https://github.com/MetaMask/core/pull/3744/commits/e80e936491c2b8a6aef5092c1192da9e7a4c0970